File: apps/plant_lab/lib/core/config/app_variant.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../access/access_context.dart';

enum AppVariant {
  plant,
  lab;

  String get code => switch (this) {
    AppVariant.plant => 'PLANT',
    AppVariant.lab => 'LAB',
  };

  String get appName => switch (this) {
    AppVariant.plant => 'Advanta Plant',
    AppVariant.lab => 'Advanta Lab',
  };

  String get moduleName => switch (this) {
    AppVariant.plant => 'Plant',
    AppVariant.lab => 'Lab',
  };

  String get tagline => switch (this) {
    AppVariant.plant => 'Plant Quality Operations',
    AppVariant.lab => 'Laboratory Quality Operations',
  };

  bool get isPlant => this == AppVariant.plant;
  bool get isLab => this == AppVariant.lab;

  bool allows(AccessContext access) => switch (this) {
    AppVariant.plant => access.canAccessPlant,
    AppVariant.lab => access.canAccessLab,
  };
}

final appVariantProvider = Provider<AppVariant>((ref) {
  throw StateError('An explicit Plant or Lab application variant is required.');
});

File: apps/admin/src/lib/appVariant.ts
import 'server-only';

export type AppVariantCode = 'PLANT' | 'LAB';

export type AppVariantConfig = {
  code: AppVariantCode;
  appName: string;
  moduleName: string;
  tagline: string;
  logoPath: string;
  accessCapability: 'can_access_plant' | 'can_access_lab';
};

const variants: Record<AppVariantCode, AppVariantConfig> = {
  PLANT: {
    code: 'PLANT',
    appName: 'Advanta Plant',
    moduleName: 'Plant',
    tagline: 'Plant Quality Operations',
    logoPath: '/logo_plant.png',
    accessCapability: 'can_access_plant',
  },
  LAB: {
    code: 'LAB',
    appName: 'Advanta Lab',
    moduleName: 'Lab',
    tagline: 'Laboratory Quality Operations',
    logoPath: '/logo_lab.png',
    accessCapability: 'can_access_lab',
  },
};

export function getAppVariant(): AppVariantConfig {
  const configuredVariant = process.env.PLANT_LAB_APP_VARIANT
    ?.trim()
    .toUpperCase();

  if (configuredVariant === 'PLANT' || configuredVariant === 'LAB') {
    return variants[configuredVariant];
  }

  throw new Error(
    'PLANT_LAB_APP_VARIANT must be set to PLANT or LAB for this deployment.',
  );
}
